### DISCUSSION TITLE
The useClipboard and useDisclosure hooks should return arrays

### DISCUSSION THREAD

**Discussion by Samin100** [chakra-ui/chakra-ui#3489] [Ideas]
# 🚀 Feature request
The `useClipboard` and `useDisclosure` hooks should return arrays rather than objects to maintain consistency with the React standard library and other existing Chakra UI hooks. If a hook returns an array, you can name the variables whatever you want. If it returns an object, then destructuring the names must be the same as in the returned object, meaning you could use a hook only once unless you rename the values.

## 🧱 Problem Statement / Justification
Right now, some Chakra hooks such as `useClipboard` and `useDisclosure` return an object rather than array which makes using the same hook more than once tedious, as you're forced to [rename variables upon destructing](https://wesbos.com/destructuring-renaming). This is different than the hooks built into React, which don't require ES6's renaming upon destructing.

## ✅ Proposed solution or API

Existing React hooks such as `useState` return an array, which are typically destructed, allowing us to quickly use the same hook more than once easily. 

Chakra's `useClipboard` and `useDisclosure` hooks don't match that pattern because they return objects. Converting them to return arrays would make using the same hook more than once simpler. Here's a real-world example:

```
# currently useClipboard currently returns an object, 
# so each value must be renamed if the hook is used more than once
const {hasCopied: hasCopiedLink, onCopy: onCopyLink} = useClipboard(link)
const {hasCopied: hasCopiedEmail, onCopy: onCopyEmail} = useClipboard(email)

# ideally all hooks should return arrays, making renaming unnecessary
const [hasCopiedLink, onCopyLink] = useClipboard(link)
const [hasCopiedEmail, onCopyEmail] = useClipboard(email)
```
## ↩️ Alternatives
The complexity of changing an existing hook's return value might not outweigh the benefits of hooks returning arrays.

## 📝 Additional Information

Additional discussion here:
[1] https://www.reddit.com/r/reactjs/comments/drltaq/why_do_hooks_return_an_array_rather_an_object/

**Comment by dodas** [comment]
Hey @Samin100.

I think returning an array (tuple) is only appropriate when a hook returns **2** elements/fields.
The [`useClipboard` hook returns **3** fields](https://chakra-ui.com/docs/hooks/use-clipboard#return-value) and the [`useDisclosure` returns **4** fields](https://chakra-ui.com/docs/hooks/use-disclosure#return-value).
Destructing from a tuple of 4 elements can get tedious really fast, if you for example need only the first and last element.

**Comment by Maadtin** [comment]
What I Usually do If I'm using more than one of these hooks It's that instead of destructuring I just use the whole object and access them by their property

**Comment by JClackett** [comment]
I think the current API is great as it is!

```ts
const copyEmail = useClipBoard(email)
const copyLink = useClipBoard(email)

// use
copyEmail.onCopy()
copyLink.onCopy()
```

I use this all the time with useDisclosure() and it means composing Modals is really easy as I can just do:

```ts
const editModalProps = useDisclosure()
const newModalProps = useDisclosure()

<Modal {...editModalProps} />
<Modal {...newModalProps} />
```

And if that's a custom modal I can spread those props onto the underlying Chakra modal and I that's it!
